[LSP] Add `textDocument/references` & `prepareRename`, Enhancements for resolving symbols

Booksbaum opened the pull request:
Started as just adding support for `textDocument/references` (godotengine/godot-proposals#3687) -- but led to a bunch of fixes & enhancements regarding what symbol is under the cursor/at a certain location (-> better results for: where are usages of a certain symbol). Which improves results for other LSP methods too.

Note: with the exception of one small addition to `./modules/gdscript/gdscript_editor.cpp`, all changes are isolated to just LSP (inside `./modules/gdscript/language_server/`)


* Fix godotengine/godot-proposals#3687: Add: `textDocument/references` ("Find All References" in VSCode):  
  ![find_references_class](https://github.com/godotengine/godot/assets/15612932/c776cc8e-f1a4-43bf-8f6b-e83e5c55d126)
  * works with native things too:  
    ![find_references_vector2](https://github.com/godotengine/godot/assets/15612932/721618a0-beca-4678-b91a-61b7b58968b8)
* Enhancement for `textDocument/rename`: prevent renaming of native or external symbols
* Add: `textDocument/prepareRename`
  * -> if rename at location not valid, VSCode prevents renaming (-> no Rename popup to change name)
    * prev:  
      ![rename_int_old](https://github.com/godotengine/godot/assets/15612932/0cdf445a-20e7-4588-b88d-d2ef6ea4fe08)
    * now:  
      ![rename_int_new](https://github.com/godotengine/godot/assets/15612932/caf3aba7-9f2b-43b7-a307-1f681c3bdcd3)
* Enhancement for `rename`/`references`: Don't search through other script files when symbol is local to current file
* Fix: `rename`ing a parameter at usage renames function instead of parameter
* Enhancement: `textDocument/documentSymbol` now returns `DocumentSymbol`s instead of `SymbolInformation`s  
  (`SymbolInformation` is marked as [deprecated](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolInformation) in LSP specs)
  * -> Enhancement: In VSCode Outline (and breadcrumb above source): Hierarchical symbols instead of flat list:
    * prev:   
     ![outline_old](https://github.com/godotengine/godot/assets/15612932/504c59f8-a93a-4428-bef3-85f4759e3bd8)  
      (hierarchy is auto-created by VSCode)
    * now:  
      ![outline_new](https://github.com/godotengine/godot/assets/15612932/e7c2936a-6550-4f73-b190-8564d0a1e9d2)
* Enhancement: In `textDocument/documentSymbol`: return only public symbols, not local ones (like parameters or variables inside a `func`)
    * see images above. For example:
      * prev: all variables inside `do_stuff` were listed (`value`, `desc`)
      * now: no local variables
* Fix: Unrelated tooltip on punctuations, parens, operators
  * prev:  
    ![tooltip_colon_old](https://github.com/godotengine/godot/assets/15612932/ca42eea5-e5de-47ef-81c2-1390b717d6e9)  
    -> some unrelated fallback
  * now:  
   ![tooltip_colon_new](https://github.com/godotengine/godot/assets/15612932/5dac9add-d2e1-4c0b-945c-51d99d425f65)  
    -> no tooltip
  (another example: see images in next fix)
* Fix/Enhancement: Commands now recognize identifier when cursor at end of identifier  
  (mentioned in https://github.com/godotengine/godot/issues/65563#issuecomment-1241695785)  
  (Cursor is between `r` and `:` )
  * prev:  
    ![tooltip_end_identifier_old](https://github.com/godotengine/godot/assets/15612932/06424e3d-42c1-4bd5-914b-88193279fc7c)
  * now:   
    ![tooltip_end_identifier_new](https://github.com/godotengine/godot/assets/15612932/0ae733d7-3809-4bd7-81d4-0d9ba8d23b5d)
    * Note: That's great with line-cursor, but a bit strange for block-cursor or mouse: As shown in example above (block-cursor): the character after identifier now "belongs" to identifier.  
      That's because LSP uses cursor position that are BETWEEN characters, while block-cursor (and mouse) is in practice on a certain character/column.  
* Fix: Incorrect column/character (Positions & Ranges) when line contains tabs  
  * -> Fixes godotengine/godot-vscode-plugin#464
  * Reason: tab in LSP is single character (`\t`), while GDScript Parser expands these based on `text_editor/behavior/indent/size`. Previously that wasn't handled, while now there's a conversion between these two if necessary.
* Enhancement/Fix: incorrect, imprecise or missing ranges and selection ranges (in `DocumentSymbol` -> ranges VSCode uses for selection and cursor movement) 
* Fix: Symbol is resolved to first declaration of symbol with same name
  * prev:  
    ![hover_variable_first_old](https://github.com/godotengine/godot/assets/15612932/e9b03952-bb08-4444-9e4d-c6c9e637f58d)  
    -> shows tooltip for `int` `value` in other `func`, but hover is on `string` `value`
  * now:  
![hover_variable_first_new](https://github.com/godotengine/godot/assets/15612932/52e32ca1-1b1b-4e2d-9a46-ce1826491874)
* Fix godotengine/godot-vscode-plugin#448: Rename sometimes does not work properly
  * Note: Not all cases are fixed. For example: symbol name inside string (or comment) still gets sometimes renamed too  
    -> Only partial fix for godotengine/godot-vscode-plugin#361 / #65563

<br/>

* And in general several Fixes and Enhancements regarding [`resolve_symbol`](https://github.com/godotengine/godot/blob/031aa99d313ceec7cdb8263957d22cae10de0b95/modules/gdscript/language_server/gdscript_workspace.h#L91) (what is under the cursor, where is it defined).  
  The notable fixes (or the ones I remember, or are easy to showcase) are already mentioned above. This here just as extra bullet point to provide some general Notes:
  * Note: with one small exception these fixes are all just for LSP `GDScriptWorkspace::resolve_symbol`, NOT `GDScriptLanguage::lookup_code` (-> for Godot editor too). LSP already had couple of special handlings for things `lookup_code` failed or didn't return all needed data (columns) and I expanded on these.  
    Long-term solution and proper fixes would probably be to handle all things in `lookup_code` (and remove special cases from LSP) -- but also more difficult and might need changes in related classes (like GDScriptParser) too.  
    But this also means: I just added or expanded special handling for "simple" cases and just for LSP.  
  * Note: there are some cases not handled by LSP, or even issues introduced by LSP (as far as I know already existed prior to this PR).  
    For example:
    * words inside strings & comments get resolved as symbols:
      ```gdscript
      var value = 42
      # comment containing value
      ```
      -> `value` in comments gets resolved to usage of property `symbol`  
      -> results in: tooltip for hover, Find References returns occurrence in comment too, rename renames `value` in comment too.
      * issue is in LSP: if `lookup_code` returns nothing, it tries to find member with same name as word under cursor. Often right, but here it's wrong
    * shadowed variable inside a variable initializer (rhs) doesn't always get resolved to correct variable:
      ```gdscript
      var value = 42
      func f():
        var value = value + 42
        #           ^^^^^
      ```
      -> usage of `value` gets resolve to `value` defined in same line instead of member `value`
      * I think issue is in `GDScriptLanguage::lookup_code`: doesn't consider rhs as part of outer suite
    * cannot resolve Element of Named Enum
      ```gdscript
      enum Named { Alpha, Beta }
      var v := Named.Alpha
      ```
      -> no tooltip or GoTo Definition for `Alpha`
      * I think issue is in `GDScriptLanguage::lookup_code` (-> fails to resolve, didn't look closer)


<br/>
<br/>

* Add some tests for LSP (test suite: `[Modules][GDScript][LSP]`)
  * Note: Not a complete cover of LSP methods -- just some of the things I wanted reproducible tests (`resolve_symbol` & converting Positions between LSP & Godot)
  * Note: To check symbols I use comment annotations in gdscript examples (see scripts in `./modules/gdscript/tests/lsp_project/`) ... which led to A LOT of tests ... which take some time: LSP tests are slower than gdscript tests:
    ```pwsh
    > Measure-Command { .\bin\godot.windows.editor.dev.x86_64.console.exe --test --test-suite="[Modules][GDScript]" | Out-Default }

    [...]
    [doctest] test cases:   1 |   1 passed | 0 failed | 799 skipped
    [doctest] assertions: 437 | 437 passed | 0 failed |
    [doctest] Status: SUCCESS!
    [...]
    TotalSeconds      : 3.2028532



    > Measure-Command { .\bin\godot.windows.editor.dev.x86_64.console.exe --test --test-suite="[Modules][GDScript][LSP]" | Out-Default }

    [...]
    [doctest] test cases:     2 |     2 passed | 0 failed | 798 skipped
    [doctest] assertions: 55742 | 55742 passed | 0 failed |
    [doctest] Status: SUCCESS!
    [...]
    TotalSeconds      : 19.4215479
    ```
    If that's too many checks/too slow I can reduce them.
  * Note: some tests are uncommented (in gdscript: with `#!`). These are some of the not handled cases mentioned above.

Booksbaum committed:
Add `==` for lsp Position and Range

Booksbaum committed:
Remove `SymbolInformation`

is obsolete

Instead `textDocument/resolveSymbols` now returns `DocumentSymbol`.  
Practical difference in VSCode: in breadcrumb navigation bar at top of source file:
Instead of a flat list of all elements its now a hierarchical tree.

Booksbaum committed:
Add & Enable `references`

Note: just stub -- doesn't return anything yet

Booksbaum committed:
Add `ReferenceParams`

Booksbaum committed:
Note if a symbol is local to a script file

Booksbaum committed:
Implement `references`

Note: `includeDeclaration` isn't handled yet (decl is always included)

Add `find_usages_in_file` & `find_all_usages`

Booksbaum committed:
Use `find usages` methods for `rename`

Enhancement: Don't look through other files when renaming local symbol

Booksbaum committed:
Emit no `children` element when there are no `children`

Booksbaum committed:
Fix: `resolve_symbol` resolves parameter to its function

Example:
```gdscript
func do_stuff(arg: int) -> int:
  arg += 2
  return arg
```
-> `resolve_symbol` on `arg` (in `arg += 2`) returns `do_stuff` (the function), NOT `arg` (the parameter)

Reason: `resolve_symbol` searches for the first symbol in the definition line -- which is the function, NOT the parameter.

Effects of this fix:
* Tooltip for `arg` usage now correctly displays parameter info instead of function info
* `Go to definition` now locates parameter instead of function
* `Find all references` now finds references of parameter instead of references of function
* `Rename` now renames parameter instead of function
* **Regression**: resolves to first symbol declaration with that name. If there's another variable/parameter with same name defined before in script it resolves to that symbol. This issue was already present before with things other than parameters, but this commit "introduces" the issue to parameter.
  * Example:
    ```gdscript
    func a(arg: int):
      pass
    func b(arg: int):
      arg += 1
    ```
    -> `arg` (in `arg += 1`) resolve to `arg` in `a`
  * Reason: prev: parameter was resolved to its function. Now its correctly resolved and as result: handled like variable.

Note: I now require the definition symbol to have the same name as usage name. (In example above: symbol must be named `arg`).
A better solution would be to adjust `GDScriptLanguage::lookup_code` to return not just the line number (`location`), but also the column (or even better: complete range -- or maybe even node?). But that's out of scope of this PR. And additionally: would require deeper changes (`lookup_code` & `LookupResult` isn't just local to GDScript, but for all `ScriptLanguage`s).

Booksbaum committed:
Format code

Booksbaum committed:
Change `prepareRename` to return `range`

prev: return `defaultBehavior: true`

Reason for change:
`defaultBehavior` doesn't work with current `godot-vscode-plugin`:
Uses old `vscode-languageclient` -- and updating requires significant more changes than just updating the package.
-> current `godot-vscode-plugin` (Version `1.3.1`) works with `range`, but not `defaultBehavior`

Fix/Enhancement: `ExtendGDScriptParser::get_identifier_under_position` returns incorrect or unclear offset
Though I don't think that had any real consequences -- might also just me misunderstanding how exactly it should be used.
To better document and clarify: `get_identifier_under_position` now returns a `lsp::Range` which is:
* zero based
* start: inclusive
* end: exclusive

Example:
```gdscript
[...]
var member := 2
```
Hover on `b` in `member`: (pos: `{ line: 3, character: 7}`)
* prev: offset (`Vector2i`): `(-4, 2)` -> range in `line: 3`: from `char: 3` to `char: 9`
  * `char: 3` exclusive? or 1-based & inclusive?
  * `char: 9`: inclusive? or 1-based & exclusive?
* now: LSP range `{ "start": { "line": 3, "character": 4 }, "end": { "line": 3, "character": 10 } }`
  * `char: 4`: inclusive
  * `char 10`: exclusive

Booksbaum committed:
Add `textDocument/prepareRename`

prevents renaming of invalid rename locations, like keywords or builtins

Effect in VSCode: rename box doesn't open when rename not valid.

Booksbaum committed:
Fix: cannot find reference/rename variable declared in other script

Example:
* `a.gd`:
  ```gdscript
  extends Node

  var member = 42
  ```
* `b.gd`:
  ```gdscript
  extends Node

  const A = preload("res://a.gd")
  func test(a: A) -> void:
    var v = a.member
  ```

-> `Find all references` (or `rename`) on `member` inside:
* `a.gd` -> only find `member` inside `a.gd`
* `b.gd` -> doesn't find any references

Issue was: variable was always treated as local

Booksbaum committed:
Add LSP test suite (`[Modules][GDScript][LSP]`)

run with `--test --test-suite="[Modules][GDScript][LSP]"`

Note: Currently doesn't pass

Note: doesn't cover much. For further testing. Currently focus on `resolve_symbol`

Booksbaum committed:
Change LSP tests to use inline markers

Ranges are now directly marked inside gdscript.
Every Range can be annotated with a name and a ref

Rewrite LSP tests from prev commit to use inline markers
Note: still failing: range of symbol spans more than just identifier

Booksbaum committed:
Fix/Adjust ranges in `DocumentSymbol`

There are two ranges:
* `range`: covers the complete node
* `selectionRange`: covers just the identifier

Note: Fixes some tests, but not all:
`resolve_symbol` for `test`s is still failing.
I think because GDScript AST doesn't keep tabs as characters but converts them to columns?

Booksbaum committed:
Enhancement: `get_identifier_under_position` (& `resolve_symbol`) now uses cursor position (instead of character position)

Consequence: Commands now recognize identifier when cursor at end of identifier

Example: (`|`: cursor, `^`: position)
```gdscript
var member| := my_func|(my_var|)
```
* prev: position was used to determine identifier
        -> no valid identifier under position, but instead space or parens
        -> rename didn't work
* now: cursor is BETWEEN chars 
       -> position is adjusted when at end of identifier
       -> rename does work



Note:  
Commands via mouse on character AFTER identifier (`^` in example above) now apply to identifier. That's just how LSP work: no difference between cursor (between chars) & mouse interaction (on char). `Position` in LSP is specified as "between two character" -- which is what this commit does.

Other extensions show same behaviour. For example: `rust-analyzer` allows commands on characters after identifier too.  
Though not all extensions: `C/C++` extension allows rename & find references on char after identifier, but doesn't show hover.

Booksbaum committed:
Enhancement: handle `includeDeclaration` in `textDocument/references`

Booksbaum committed:
Add rename-is-valid check to `textDocument/rename`

Same check as in `textDocument/prepareRename`

Booksbaum committed:
Fix: Incorrect Position when `\t` somewhere before column

Godot expands `\t` with `text_editor/behavior/indent/size`, while in LSP `\t` is just a single character.
-> Requires conversion between LSP and Godot

Note: currently only parsing to `DocumentSymbols`s converts Godot to LSP.
That should cover most cases -- especially `resolve_symbol`.  
But there might be still some unhandled cases -> not yet checked.

Note: requires more tests -- currently only simple test with single tab

Booksbaum committed:
Fix: Char after Cursor gets lost while marking Cursor in String

Booksbaum committed:
Simplify `find_all_usages`

Addendum to prev commit message:
prev commit seems to cover all cases which require conversion between LSP and Godot.
Currently there seem to be only two main bridges between GDScriptParser & ExtendGDScriptParser:
* parsing in ExtendGDScriptParser -> transforms GDScript AST (Godot Positions) into `DocumentSymbol` (LSP positions). That's then used inside `resolve_symbol` (-> LSP positions)
* `GDScriptLanguage::get_text_for_completion` && `lookup_code` -> take source code with cursor marked in string instead of extra passed position -> no explicit conversion necessary

Booksbaum committed:
Add: Check for errors in GDScripts used in tests

Booksbaum committed:
Add more tests for ranges & indentation

Booksbaum committed:
Add: Check for all names unique in tests

Booksbaum committed:
Handle lambda in fields

Merge lambdas with parent variable

Booksbaum committed:
Use docs from GDScriptParser instead of extra doc parser

Booksbaum committed:
Fix: resolve_symbol resolves to wrong symbol in certain cases

Handled here:
* in getter/setter
* in lambdas inside functions
* parameters in signals
* in different scopes
* inside match
* when script contains multiple variables with same name

Though not all cases are fixed. For example:
* lambdas in fields
* match cases
* shadowed variables  
  Though: shadowing is discouraged in gdscript

Booksbaum committed:
Add: Test front and back of identifiers too

instead of just middle

Note: greatly increases test duration. GDScript Tests are now faster than LSP tests...
Too many tests? tone down?

Booksbaum committed:
Rename SUBCASEs

Booksbaum committed:
Enhancement: Only return non-local `DocumentSymbol`s

in `textDocument/resolveSymbol`.
VSCode shows result as breadcrumbs & navigation at top of source.
With local `DocumentSymbol`: includes parameters & variables local to `func`s
-> quite crowded & distracting clutter
-> only emit public `DocumentSymbol`s (classes, variables, signals, functions, ...)

(same behaviour as in other LSP implementations)

Booksbaum committed:
Add some safety check & tests for lsp Position to Godot Position

Booksbaum committed:
Fix: Cannot resolve variable decl when shadowing variable with same name

Example:
```gdscript
var value := 42

func f():
  var value = value + 42
```
2nd `value` (decl in `f`) was resolved to 1st `value` (decl in class)

Note:
Still faulty: `value` in rhs cannot be resolved to correct `value`:
Gets resolved to local `value` (decl in `f`) instead of public `value` (decl in class).
-> some tests are uncommented (with `#!`)

I *think* that's because `GDScriptLanguage::lookup_code` just checks if a suite (scope) has a local variable matching checked identifier, but doesn't check if usage is in initializer (-> identifier belongs to outer scope).

Surprisingly the issue doesn't emerge for all declarations -- but only some. For example: `for value in value:` seems to work. I *think* that's because `GDScriptLanguage::lookup_code` cannot resolve the loop variable (1st `value`) (which is then resolved via some special resolving for local members in LSP code), and as such there's no "conflict" for `GDScriptLanguage::lookup_code` with anything declared in same line.

Booksbaum committed:
Add test for ctor

Booksbaum committed:
Partial Fix: Cannot resolve Values in Named Enum

Note: Only fixed in definition -- NOT usage (`Named.MyValue`)

Reason:
`GDScriptLanguage::lookup_code` cannot resolve `Named.MyValue`.
Going deeper: `GDScriptParser::DataType` doesn't contain source for enums -- only Name & Values -> unlike classes enum locations cannot be retrieved (see `gdscript_editor.cpp` -> `_lookup_symbol_from_base`).
Changing that would require something like adding `EnumName *enum_type` (similar to `class_type`) and adjusting usage locations.
That's not done here!

Note:
Things like Hover still work on `Named.MyValue`: unlike Find References and Rename, (when smart resolve enabled) Hover tries to find related symbols by name -> finds ALL symbols named `MyValue` in all scripts

Note:
above indirectly reveals an additional shortcoming of `GDScriptWorkspace::get_member_symbol` & `get_local_symbol`: We're only looking at identifier under cursor -- but not prev identifiers: `something.whatever.other_thing.value` -> with cursor on `value` -> only search for any variable/member matching `value`. Might resolve to something other than on type of `other_thing`.
I think that's acceptable: `GDScriptLanguage::lookup_code` should resolve the symbol (and DOES handle identifier chains) and we're only using `get_xxx_symbol` as fallbacks.
In cases that might be the wrong resolution: should fix `lookup_code`, and such not passing it on to fallbacks.

Note:
In tests (`enums.gd`), failing tests are uncommented with `#!` -> don't get executed, but are still marked in test scripts.

Booksbaum committed:
Fix: `GDScriptLanguage::lookup_code` cannot resolve inner class

Example:
```gdscript
extends Node

class Inner1:
  func my_func():
    pass

func f():
  var inner1 = Inner1.new()
  var value1 = inner1.my_func()
```
`lookup_code` with cursor in `Inner1` in `Inner1.new()` cannot resolve to `Inner1` class.
That extends to nested resolves like cursor in `my_func` in last line -> to resolve `my_func` it must resolve `inner1` to `Inner1` -- which it can't -> cannot resolve `my_func`

Booksbaum committed:
Enhancement: Run following tests if a `resolve_symbol` returns null

-> `Check` instead of `Require`

Booksbaum committed:
Add more nested class tests

Note: `extends MyClass` in nested class inside nested class fails and is uncommented!

Booksbaum committed:
Inline `==` for Position & Range

Remove old TODOs

Booksbaum committed:
Remove unused symbol

Booksbaum committed:
Note about `resolve_symbol` inside comments & strings

The way `resolve_symbol` currently works, words inside comments & strings get resolved as symbols (at least sometimes):
```gdscript
var value = 42

```
-> `value` inside comments gets recognized as usage of property `value`.

That's how LSP specific resolving currently works (already prior to this PR) when `GDScriptLanguage::lookup_code` doesn't find anything: Get identifier under cursor and then check if there's a matching member.
This happens when `lookup_code` incorrectly resolves nothing (for example: often the case for declarations of local variables & parameters) -- after all that's the reason for LSP special handling. But also when `lookup_code` correctly resolves nothing (like in string & comments).

That means: `Find All References`, `Rename` and others includes matches inside comments & strings -- which might be great (if mentioning of current symbol (-> tags)) ... or bad (if wrong symbol ... or just common word with name of existing symbol...).

A long-term solution and real fix would probably be: handle all cases in `lookup_code` (GDScript) and remove special cases for LSP.
But I didn't do that for this PR (with one exception because that was trivial) but only adjusted and expanded existing LSP special cases.

-----------------

Note: Nothing changed with this commit; it's just to document still existing special behaviour!

Booksbaum committed:
Fix: properties with extra get/set functions throw exception in `ExtendGDScriptParser::parse_function_symbol``

Booksbaum committed:
Fix: in tests on Windows: unescaped `:` (`C:/...`) in `root_uri`

Note: had no consequences as current tests don't use URI.

Booksbaum committed:
Cleanup

Booksbaum committed:
Fix: Range for GoTo Definition/Declaration spans getter/setter

Resulted in: In VSCode: in getter/setter GoTo Def/Decl on symbol brought up "Find All References" (because: inside Def/Decl range -> VSCode automatically calls Find References).

Issue was: `range` (covers full node) instead of `selectionRange` (covers only identifier)

AThousandShips commented:
Does this completely fix all the cases you have linked? If not please make that clear and do not link those issues as solved

Booksbaum commented:
The ones with one of the Github keywords should be fixed.
Maybe with slight asterisk: https://github.com/godotengine/godot-proposals/issues/3687 I *think* it uses LSP method `textDocument/references`, but I don't know how `vim.lsp.buf.references()` is actual implemented.

and maybe one extra note for https://github.com/godotengine/godot-vscode-plugin/issues/448 : as noted in description above only the exact example is fixed, while other renames aren't. but these have extra issues (https://github.com/godotengine/godot-vscode-plugin/issues/361, https://github.com/godotengine/godot/issues/65563) (also noted in description above, but carefully worded to not trigger the Github "Fix" keyword)

AThousandShips commented on modules/gdscript/language_server/gdscript_extend_parser.cpp:
```diff
@@ -32,9 +32,91 @@
+lsp::Position GodotPosition::to_lsp(const Vector<String> &p_lines) const {
+	lsp::Position res;
+
+	// special case: `line = 0` -> root class (range covers everything)
```
```suggestion
	// Special case: `line = 0` -> root class (range covers everything).
```
As per the [comment style guide](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide), same elsewhere

AThousandShips commented on modules/gdscript/language_server/gdscript_extend_parser.h:
```diff
@@ -50,6 +53,63 @@
+	int line;
+	int column;
+
+	GodotPosition(int line, int column) :
```
```suggestion
	GodotPosition(int p_line, int p_column) :
```
Use `p_` prefix for function parameters, to prevent shadowing

AThousandShips commented on modules/gdscript/tests/test_lsp.h:
```diff
@@ -0,0 +1,462 @@
+#include "core/os/os.h"
+#include "editor/editor_help.h"
+#include "editor/editor_node.h"
+#include "gdscript/gdscript_analyzer.h"
```
```suggestion
#include "modules/gdscript/gdscript_analyzer.h"
```

Booksbaum replied:
I changed it to relative path (because we're inside gdscript module). But `regex` needed a leading `modules`

Booksbaum replied:
It's how it's done in other tests.

And with "because inside gdscript module": I meant it makes logical sense because we're inside the gdscript module -- so it makes sense to use relative path like for other gdscript headers inside the test files.

Compiling works (at least for me) in all cases: with `modules`, without `modules` & relative path

AThousandShips replied:
It failed in the CI, that's why I commented, change back and you can see

Booksbaum committed:
Adjust comments to upper case sentences

Booksbaum committed:
Use `p_` prefix for parameters

AThousandShips commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -112,6 +120,27 @@ struct Range {
+		return start == p_other.start && end == p_other.end;
+	}
+
+	bool contains(const Position p_pos) const {
```
Any reason this isn't passed by reference here?

Booksbaum replied:
Not really. But neither any reason to be passed by reference: `Position` is very small (and copying is most likely faster -- probably what the compiler does anyway?)

AThousandShips replied:
We do it elsewhere for it, so uniformity, and it doesn't cost anything to do so

Booksbaum committed:
Adjust includes

AThousandShips commented on modules/gdscript/language_server/gdscript_extend_parser.cpp:
```diff
@@ -492,7 +623,7 @@ String ExtendGDScriptParser::get_text_for_completion(const lsp::Position &p_curs
 		if (i == p_cursor.line) {
 			longthing += lines[i].substr(0, p_cursor.character);
-			longthing += String::chr(0xFFFF); //not unicode, represents the cursor
+			longthing += String::chr(0xFFFF); //Not unicode, represents the cursor.
```
```suggestion
			longthing += String::chr(0xFFFF); // Not unicode, represents the cursor.
```
Though not needed to change code that you haven't touched

Booksbaum committed:
Space after `//`

Booksbaum committed:
Pass by reference

AThousandShips commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -112,6 +120,27 @@ struct Range {
+	}
+
+	bool contains(const Position &p_pos) const {
+		// inside line range
```
```suggestion
		// Inside line range.
```
Here as well

Booksbaum committed:
Comments

Booksbaum committed:
More comments

Booksbaum committed:
include `gdscript_analyzer` with `modules`

Booksbaum committed:
Add default case

Booksbaum committed:
Fix: missing `break`

Booksbaum committed:
Consume `GodotPosition` in `GodotRange` ctor

AThousandShips commented:
In the future please do not make multiple commits in such rapid succession, currently it won't fire CI because you're a first time contributor but in the future it will

Booksbaum commented:
Ok, reading files fails.

(`path="res//local_variables.gd"`)  
Why does `FileAccess::get_file_as_string(path, &err)` fail? Project should be initialized to correct directory (`root = "modules/gdscript/tests/lsp_project/"`)

And additional: why does it work locally?


-------------

difference I found: Godot Project Settings [root path](https://github.com/Booksbaum/godot/blob/83cc3a8f970a33d0720ee2d8e7cee513516227dd/modules/gdscript/tests/test_lsp.h#L80) isn't with trailing `/` (while gdscript tests use trailing `/`) ... but internal GDScript uses only its `DirAccess` etc. ... so shouldn't make any difference.

dalexeev commented:
@Booksbaum If the status of this PR is Work In Progress / Early Preview, then you can mark it as draft for now, to make it more obvious to contributors. And mark it as ready for review later, when you decide that the moment has come. Thanks!

Booksbaum commented:
Error is: `ProjectSettings::resource_path` is still from gdscript tests. `ProjectSettings::setup` doesn't change `resource_path` when called again (and maybe even more state remains?)  
-> lsp tests still try to load resource from gdscript tests


Which explains why it worked locally: I either always filtered to test-suite `[Modules][GDScript]` or `[Modules][GDScript][LSP]` -- but never both together (`[Modules][GDScript]*` -> fails)




But `ProjectSettings` doesn't have a method to reset its state ... hmpf...
In here: https://github.com/godotengine/godot/blob/031aa99d313ceec7cdb8263957d22cae10de0b95/core/config/project_settings.cpp#L486-L619 -> two `OS::get_singleton()->get_resource_dir().is_empty()` checks which use old resource_path.

Simple fix seems to be manually reset `resource_path`. Then tests work (at least local). Is it ok to expose [`resource_path`](https://github.com/godotengine/godot/blob/031aa99d313ceec7cdb8263957d22cae10de0b95/core/config/project_settings.h#L95) for test builds (or a reset method only available for test builds ... though other state remains untouched :/ )? 

<br/>
<br/>

------------------------

<br/>
<br/>


> If the status of this PR is Work In Progress / Early Preview, then you can mark it as draft for now, to make it more obvious to contributors. And mark it as ready for review later, when you decide that the moment has come. Thanks!

Wasn't planned as such :D ..... :(

But marked as WIP for now -> enough for today

Booksbaum committed:
`finish_language` after tests

Note:
Running `[Modules][GDScript]` and `[Modules][GDScript][LSP]` together (`[Modules][GDScript]*`) fails because it cannot load resources of the later test.
Issue: `ProjectSettings::resource_path` doesn't get reset inside `ProjectSettings::setup` (`_setup`) -- but instead if `resource_path` is already set it's reused -> LSP tests use `resource_path` of GDScript tests -> loading LSP resources fails

Booksbaum committed:
Fix: LSP tests fail when run together with GDScript tests

Issue was:
`ProjectSettings` cannot be reset, but uses old state (especially `resource_path`) when `setup` gets called.
-> Cannot change workspace (from `modules/gdscript/tests/scripts/` to `module/gdscript/tests/lsp/`)
-> Loading resources (script files) in LSP tests fails (because tries to read from old gdscript folder)

LSP scripts are now inside GDScript tests project, and GDScript tests & LSP tests now share a common scripts directory.
LSP scripts are inside `lsp` sub-directory.

Booksbaum committed:
Guard LSP tests behind `TOOLS_ENABLED`

Required for LSP server

Booksbaum commented:
Tests should now run.

I moved LSP test scripts into same workspace as other GDScript test scripts -> both test suites use same `ProjectSettings` state.

Booksbaum commented:
Yay...Unit tests succeeded

But:
```
ranlib: /home/runner/work/godot/godot/godot-cpp/bin/libgodot-cpp.linux.template_debug.dev.x86_64.a: No space left on device
```
in `Build godot-cpp test extension`  
Though seems to be a general CI run error and not directly caused by this PR? I don't think this PR adds so much to waste extreme amount of disk space...

AThousandShips commented:
Yes unrelated to this PR

ryanabx commented:
Great language server changes! Looking forward to this going to master :)

ryanabx commented:
I'd like to take this on and fix the unsuccessful check if possible.

> Yay...Unit tests succeeded
> 
> But:
> 
> ```
> ranlib: /home/runner/work/godot/godot/godot-cpp/bin/libgodot-cpp.linux.template_debug.dev.x86_64.a: No space left on device
> ```
> 
> in `Build godot-cpp test extension` Though seems to be a general CI run error and not directly caused by this PR? I don't think this PR adds so much to waste extreme amount of disk space...



> Yes unrelated to this PR

Was this ever fixed? Is this the only blocker to getting this approved?

AThousandShips commented:
This has been fixed and this needs a rebase to include the fix

ryanabx commented:
> This has been fixed and this needs a rebase to include the fix

Rebase finished! Had to fix some references to class variables that were moved to other places. The workflow checks passed on my fork, and I created a PR: #80973

Booksbaum commented:
superseeded by #80973
